doc/reference/networking/net_pkt.rst

```diff
@@ -10,6 +10,351 @@ Packet Management
+
+.. code-block:: c
+
+    net_pkt_write_u8(pkt, &foo);
+    net_pkt_write_be16(pkt, &ba);
+    net_pkt_write_be32(pkt, &bar);
+
+Logically, net_pkt's length is now 15. But if we try to read at this
+point, it will fail becaus there is nothing to read at the cursor
+where we are at in the net_pkt. It is possible, while in write mode,
+to read what has been already written by resetting the cursor of the
+net_pkt. For instance:
+
+.. code-block:: c
+
+    net_pkt_cursor_init(pkt);
+    net_pkt_read(pkt, data, 15);
+
+This will have reset the cursor of the pkt to the beginning of
+the buffer and then let to read the actual 15 bytes present. The
+cursor is then again pointing at the end of the buffer.
+
+To set a large area with the same byte, a memset function is provided:
+
+.. code-block:: c
+
+    net_pkt_memset(pkt, 0, 5);
+
+Our net_pkt has now a length of 20 bytes.
+
+Switching between modes can be achieved via
+:c:func:`net_pkt_set_overwrite` function. It is possible to switch
+mode back and forth at any time.  The net_pkt will be set to overwrite
+and its cursor reset:
+
+.. code-block:: c
+
+    net_pkt_set_overwrite(pkt, true);
+    net_pkt_cursor_init(pkt);
+
+Now the same operators can be used, but it will be limited to the
+existing data in the buffer, i.e. 20 bytes.
+
+If it is necessary to know how much space is available in the net_pkt
+call:
+
+.. code-block:: c
+
+    net_pkt_available_buffer(pkt);
+
+Or, if headers space needs to be accounted for, call:
+
+.. code-block:: c
+
+    net_pkt_available_payload_buffer(pkt, proto);
+
+If you want to place the cursor at a known position use the function
+:c:func:`net_pkt_skip`.  For example, to go after the IP header, use:
+
+.. code-block:: c
+
+    net_pkt_cursor_init(pkt);
+    net_pkt_skip(pkt, net_pkt_ip_header_len(pkt));
+
+
+Data access
+===========
+
+Though the API shown previously is rather simple, it involves always
+copying things to and from the net_pkt buffer. In many occasions, it
+is more relevant to access the information stored in the buffer
+contiguously, especially with network packets which embed headers.
+
+These headers are, most of the time, a known fixed set of bytes. It is
+then more natural to have a structure representing a certain type of
+header.  In addition to this, if it is known the header size appears
+in a contiguous area of the buffer, it will be way more efficient to
+cast the actual position in the buffer to the type of header. Either
+for reading or writing the fields of such header, accessing it
+directly will save memory.
+
+Net pkt comes with a dedicated API for this, built on top of the
+previously described API. It is able to handle both contiguous and
+non-contiguous access transparently.
+
+There are two macros used to define a data access descriptor:
+:c:macro:`NET_PKT_DATA_ACCESS_DEFINE` when it is not possible to
+tell if the data will be in a contiguous area, and
+:c:macro:`NET_PKT_DATA_ACCESS_CONTIGUOUS_DEFINE()` when
```

Comment:
this macro has () but the one above does not, just wondering if we should remove the () here